core/embed/util/image/boot_image.c

```diff
@@ -213,4 +218,72 @@ void bl_check_replace(const uint8_t *data, size_t len) {
   mpu_reconfig(mode);
 }
 
+#else
+
+bool boot_image_check(const boot_image_t *image) {
+  if (image->image_size < sizeof(boot_header_auth_t)) {
+    // Invalid image size, must be at least the size of the header
+    return false;
+  }
+
+  mpu_mode_t mode = mpu_reconfig(MPU_MODE_BOOTUPDATE);
+
+  boot_header_auth_t *cur_hdr = (boot_header_auth_t *)BOOTLOADER_START;
+  boot_header_auth_t *new_hdr = (boot_header_auth_t *)image->image_ptr;
+
+  bool diff = (cur_hdr->header_size != new_hdr->header_size) ||
+              (memcmp(cur_hdr, new_hdr, cur_hdr->header_size) != 0);
+
+  mpu_restore(mode);
+
+  return diff;
+}
+
+void boot_image_replace(const boot_image_t *image) {
+  uint32_t header_address = (uint32_t)image->image_ptr;
+
+  // Check that image is big enough to hold the header at least
+  ensure(sectrue * (image->image_size >= sizeof(boot_header_auth_t)),
+         "Bootloader image too small");
+
+  // Read bootloader header
+  const boot_header_auth_t *hdr = boot_header_auth_get(header_address);
+  ensure((hdr != NULL) * sectrue, "Invalid bootloader header");
+
+  // Check the image is big enough to hold both header and code
+  ensure(sectrue * (hdr->header_size + hdr->code_size <= image->image_size),
+         "Bootloader image too small");
+
+  // Check monotonic version
+
+  mpu_mode_t mpu_mode = mpu_reconfig(MPU_MODE_BOOTUPDATE);
+
+  const boot_header_auth_t *old_hdr = boot_header_auth_get(BOOTLOADER_START);
+
+  ensure((old_hdr != NULL) * sectrue, "Invalid current bootloader header");
```

Comment:
this is a problem!
means that if there's broken old header, the update to a fixed one will not proceed

Reply:
we can do something like read header, if succeeded check monotonic version, otherwise continue without monotonic check